Add base Ruby 2.1.5 image and support for image versions

raganw opened the pull request.

znep commented:
I'd recommend including the version in the image name (unlike the current java image) so version dependencies can be managed.

znep commented on ruby/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+  rm -rf /var/lib/apt/lists/*
+
+# skip installing gem documentation
+RUN echo 'gem: --no-rdoc --no-ri' >> "$HOME/.gemrc"
```
Can you combine these two RUNs to avoid unnecessary intermediate images, given they are really part of the same step?  Also, did you double check that $HOME is set?

raganw replied:
```
$ docker run -it socrata/base bash
root@77a226163388:/# echo $HOME
/root
root@77a226163388:/#
```

Now, if a project wants to run as the 'socrata' user that's been added to the base image, it'll need to handle setting its own gem mumbo-jumbo.

znep replied:
Cool, thx for verifying.

Presumably an app won't normally be installing gems at container runtime so
even if it runs as socrata these particular settings won't matter, right?
These gems are installed system wide?
On Jan 14, 2015 8:17 AM, "Ragan Webber" <notifications@github.com> wrote:

> In ruby/Dockerfile
> <https://github.com/socrata/shipyard/pull/7#discussion-diff-22947458>:
>
> > @@ -0,0 +1,15 @@
> > +FROM socrata/base
> > +MAINTAINER Socrata <sysadmin@socrata.com>
> > +
> > +RUN DEBIAN_FRONTEND=noninteractive apt-get -y update && \
> > +  DEBIAN_FRONTEND=noninteractive apt-get --force-yes -fuy install software-properties-common && \
> > +  DEBIAN_FRONTEND=noninteractive add-apt-repository -y ppa:brightbox/ruby-ng && \
> > +  DEBIAN_FRONTEND=noninteractive apt-get -y update && \
> > +  DEBIAN_FRONTEND=noninteractive apt-get install -y ruby2.1 && \
> > +  DEBIAN_FRONTEND=noninteractive apt-get purge -y --auto-remove software-properties-common && \
> > +  rm -rf /var/lib/apt/lists/*
> > +
> > +# skip installing gem documentation
> > +RUN echo 'gem: --no-rdoc --no-ri' >> "$HOME/.gemrc"
>
> $ docker run -it socrata/base bash
> root@77a226163388:/# echo $HOME
> /root
> root@77a226163388:/#
>
> Now, if a project wants to run as the 'socrata' user that's been added to
> the base image, it'll need to handle setting its own gem mumbo-jumbo.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/socrata/shipyard/pull/7/files#r22947458>.
>

znep commented on ruby/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+RUN DEBIAN_FRONTEND=noninteractive apt-get -y update && \
+  DEBIAN_FRONTEND=noninteractive apt-get --force-yes -fuy install software-properties-common && \
+  DEBIAN_FRONTEND=noninteractive add-apt-repository -y ppa:brightbox/ruby-ng && \
```
Can you add the repo first then do a single update and a single install?

raganw replied:
`software-properties-common` provides the `add-apt-repository` command, so, no.

raganw commented:
Pull request updated with requested changes, including support for version sub-directories. :tada:

znep commented on bin/build:
```diff
@@ -5,13 +5,34 @@ set -e
-  docker build $BUILD_OPTS -t ${NS}/$tag $DIR/$subdir
+  if [ -n "$versions" ]; then
+    for version in $versions; do
+      docker build $BUILD_OPTS -t ${NS}/$tag:$version $DIR/$subdir
```
So just making sure my understanding is correct... and if it is then this looks good.

If you build a new image with the same tag (dockerfile change, security update, etc. etc.) then you have to make sure to explicitly pull that image before building things that depend on it in order to ensure you get the updated version, right?  But once that is done, there is no deploy time problem to deal with.  And none of this really is any different having the version or not having it.

raganw replied:
Right.  What we are calling "version" here, in Docker parlance is actually "tag" (and what we're calling "tag" is the image name).  There is an implicit "latest" tag if it is not explicitly set.  But, if an image has already been cached locally, building on top of it won't fetch the latest from the the docker registry.  That would require an explicit `docker pull`.

znep commented on ruby/2.1.5/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+  DEBIAN_FRONTEND=noninteractive apt-get --force-yes -fuy install software-properties-common && \
+  DEBIAN_FRONTEND=noninteractive add-apt-repository -y ppa:brightbox/ruby-ng && \
+  DEBIAN_FRONTEND=noninteractive apt-get -y update && \
+  DEBIAN_FRONTEND=noninteractive apt-get install -y ruby2.1 && \
```
So you are installing the 2.1 package but calling it 2.1.5.  Does that make sense or should we just call it 2.1?  Won't the package potentially install 2.1.6, etc. at some point?  This reminds me of the ubuntu situation we got into where "1.9.1" was actually "1.9.3" or some such.

(not trying to be nitpicky...)

raganw replied:
Good point. I'll make that just 2.1.

znep commented:
LGTM other than my two latest comments.

raganw commented:
Okay, I've updated the version directory name and fixed an issue with the build script not building both the untagged latest and tagged version for when they exist.

znep commented:
Thanks for the explanations.
